src/satosa/backends/idpy_oidc.py

```diff
@@ -0,0 +1,148 @@
+        if not _redirect_uris:
+            raise SATOSAError("Missing path in redirect uri")
+        self.redirect_path = urlparse(_redirect_uris[0]).path
+
+    def start_auth(self, context, internal_request):
+        """
+        See super class method satosa.backends.base#start_auth
+
+        :type context: satosa.context.Context
+        :type internal_request: satosa.internal.InternalData
+        :rtype satosa.response.Redirect
+        """
+        login_url = self.client.init_authorization()
+        return Redirect(login_url)
+
+    def register_endpoints(self):
+        """
+        Creates a list of all the endpoints this backend module needs to listen to. In this case
+        it's the authentication response from the underlying OP that is redirected from the OP to
+        the proxy.
+        :rtype: Sequence[(str, Callable[[satosa.context.Context], satosa.response.Response]]
+        :return: A list that can be used to map the request to SATOSA to this endpoint.
+        """
+        url_map = []
+        url_map.append((f"^{self.redirect_path.lstrip('/')}$", self.response_endpoint))
+        return url_map
+
+    def response_endpoint(self, context, *args):
+        """
+        Handles the authentication response from the OP.
+        :type context: satosa.context.Context
+        :type args: Any
+        :rtype: satosa.response.Response
+
+        :param context: SATOSA context
+        :param args: None
+        :return:
+        """
+
+        _info = self.client.finalize(context.request)
+        self._check_error_response(_info, context)
+        userinfo = _info.get('userinfo')
+        id_token = _info.get('id_token')
+
+        if not id_token and not userinfo:
+            msg = "No id_token or userinfo, nothing to do.."
+            logline = lu.LOG_FMT.format(id=lu.get_session_id(context.state), message=msg)
+            logger.error(logline)
+            raise SATOSAAuthenticationError(context.state, "No user info available.")
+
+        all_user_claims = dict(list(userinfo.items()) + list(id_token.items()))
+        msg = "UserInfo: {}".format(all_user_claims)
+        logline = lu.LOG_FMT.format(id=lu.get_session_id(context.state), message=msg)
+        logger.debug(logline)
+        internal_resp = self._translate_response(all_user_claims, _info["issuer"])
+        return self.auth_callback_func(context, internal_resp)
+
+    def _translate_response(self, response, issuer):
+        """
+        Translates oidc response to SATOSA internal response.
+        :type response: dict[str, str]
+        :type issuer: str
+        :type subject_type: str
+        :rtype: InternalData
+
+        :param response: Dictioary with attribute name as key.
+        :param issuer: The oidc op that gave the repsonse.
+        :param subject_type: public or pairwise according to oidc standard.
+        :return: A SATOSA internal response.
+        """
+        timestamp = response["auth_time"] or datetime.utcnow()
```

Comment:
response["auth_time"] may lead to KeyError if auth_time is missing from the response.

Moreover, datatime.utcnow() would return datetime object which is not serializable. I think we need to change it to something as follows:

`
timestamp = response.get("auth_time")
 if not timestamp:
     timestamp = datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%SZ")
`

Reply:
`auth_time` is required to be there.

The datetime should be a timestamp
```suggestion
        timestamp = response["auth_time"] or int(datetime.utcnow().timestamp())
```

Reply:
I tried to authenticate via Microsoft AD and got KeyError at this line and that's why I thought may be response would not always have `auth_time`. Do you think it's an issue with MS AD? Nonetheless, I think your following suggestion:

`timestamp = response["auth_time"] or int(datetime.utcnow().timestamp())
`
should be adopted.

Reply:
I am sorry, I went back to the spec to check and indeed `auth_time` is not required. `iat` is required, so that would be a better choice. So probably just have

```suggestion
        timestamp = response["iat"]
```

but I think it is also ok to have

```suggestion
        timestamp = (
            response.get("auth_time")
            or response.get("iat")
            or int(datetime.utcnow().timestamp()
        )
```

Reply:
Yes, I agree @c00kiemon5ter  - thanks.

Reply:
We should put the right value here. I'll look into this.
Probably tomorrow.